GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1078,30 +1087,85 @@ private void UpdatePlaceholders()
         {
             this.ClearUpdatePlaceholderErrors();
 
-            List<PlaceholderListDatabase.PlaceholderData> placeholderListCopy = this.placeholderList.GetAllEntries();
+            List<PlaceholderListDatabase.PlaceholderData> placeholderFilesListCopy;
+            List<PlaceholderListDatabase.PlaceholderData> placeholderFoldersListCopy;
+            this.placeholderList.GetAllEntries(out placeholderFilesListCopy, out placeholderFoldersListCopy);
+
             EventMetadata metadata = new EventMetadata();
-            metadata.Add("Count", placeholderListCopy.Count);
+            metadata.Add("File placeholder count", placeholderFilesListCopy.Count);
+            metadata.Add("Folder placeholders count", placeholderFoldersListCopy.Count);
+
             using (ITracer activity = this.context.Tracer.StartActivity("UpdatePlaceholders", EventLevel.Informational, metadata))
             {
+                int minItemsPerThread = 10;
+                int numThreads = Math.Max(8, Environment.ProcessorCount);
+                numThreads = Math.Min(numThreads, placeholderFilesListCopy.Count / minItemsPerThread);
+                numThreads = Math.Max(numThreads, 1);
+
                 ConcurrentHashSet<string> folderPlaceholdersToKeep = new ConcurrentHashSet<string>();
-                ConcurrentBag<PlaceholderListDatabase.PlaceholderData> updatedPlaceholderList = new ConcurrentBag<PlaceholderListDatabase.PlaceholderData>();
+
+                // On platforms that expand on enumeration we need to keep the updated placeholders in a collection
+                // that supports fast searching (for ReExpandFolder)
+                ConcurrentDictionary<string, PlaceholderListDatabase.PlaceholderData> updatedPlaceholderDictionary;

```

Comment:
It's a little concerning that we're keeping two data structures for the same data. How do we guarantee that they stay in sync?

Reply:
Or is the requirement that it be one or the other? In that case, I'd add a few strategic checks that enforce that and throw an exception if it's violated.